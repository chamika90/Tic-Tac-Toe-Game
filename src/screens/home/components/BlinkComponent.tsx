import React, { ReactElement, useEffect } from 'react';
import { Animated, View } from 'react-native';

type BlinkComponentProps = {
    duration: number;
    children: ReactElement;
};
export const BlinkComponent: React.FC<BlinkComponentProps> = ({ duration, children }) => {
    const fadeAnimation = new Animated.Value(0);

    useEffect(() => {
        Animated.loop(
            Animated.sequence([
                Animated.timing(fadeAnimation, { toValue: 0, duration: duration, useNativeDriver: true }),
                Animated.timing(fadeAnimation, { toValue: 1, duration: duration, useNativeDriver: true }),
            ]),
        ).start();
    }, []);

    return (
        <View>
            <Animated.View style={{ opacity: fadeAnimation }}>{children}</Animated.View>
        </View>
    );
};
