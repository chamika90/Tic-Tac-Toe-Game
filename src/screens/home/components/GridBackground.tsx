import React from 'react';
import { View, StyleSheet } from 'react-native';
import { COLORS } from '../../../resources';

export type GridBackgroundProps = {
    cellWidth: number;
    cellContainerWidth: number;
    numberOfHorizontalLines: number;
    numberOfVerticalLines: number;
};
export const GridBackground: React.FC<GridBackgroundProps> = ({
    cellWidth,
    cellContainerWidth,
    numberOfHorizontalLines,
    numberOfVerticalLines,
}) => {
    const HorizontalLine = () => {
        return (
            <View
                testID="horizontalLine"
                style={[{ marginLeft: cellWidth, height: cellContainerWidth }, styles.horizontalLine]}
            />
        );
    };

    const VerticalLine = () => {
        return <View style={[{ marginTop: cellWidth, width: cellContainerWidth }, styles.verticalLine]} />;
    };

    const HorizontalLineContainer = () => {
        return (
            <View testID="horizontalLineContainer" style={styles.horizontalLineContainer}>
                {[...Array(numberOfHorizontalLines)].map((element, index) => (
                    <HorizontalLine key={index} />
                ))}
            </View>
        );
    };
    const VerticalLineContainer = () => {
        return (
            <View testID="verticalLineContainer" style={styles.verticalLineContainer}>
                {[...Array(numberOfVerticalLines)].map((element, index) => (
                    <VerticalLine key={index} />
                ))}
            </View>
        );
    };

    return (
        <View style={{ width: cellContainerWidth, height: cellContainerWidth }}>
            <HorizontalLineContainer />
            <VerticalLineContainer />
        </View>
    );
};

const styles = StyleSheet.create({
    horizontalLineContainer: { flexDirection: 'row', position: 'absolute' },
    verticalLineContainer: { flexDirection: 'column', position: 'absolute' },
    horizontalLine: { width: 1, backgroundColor: COLORS.WHITE },
    verticalLine: { height: 1, backgroundColor: COLORS.WHITE },
});
