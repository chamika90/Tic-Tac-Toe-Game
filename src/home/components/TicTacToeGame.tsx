import React, { useState, useEffect } from 'react';
import {
    Text,
    View,
    StyleSheet,
    Dimensions,
    TouchableWithoutFeedback,
    TouchableOpacity,
    Alert,
    Image,
} from 'react-native';
import { GridBackground } from './GridBackground';
import { BlinkComponent } from './BlinkComponent';
import { COLORS } from '../../resources';
import { isAllEqual } from '../../helpers/utils';

const windowWidth = Dimensions.get('window').width;
const cellWidth = windowWidth / 3.5;
const cellContainerWidth = cellWidth * 3;
const initialCellValues = [null, null, null, null, null, null, null, null, null];
const numberOfBoardCells = 9;
const winingRowIndexes = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [3, 5, 8],
    [0, 4, 8],
    [2, 4, 6],
];
const playerXColor = COLORS.BLUE_50;
const playerYColor = COLORS.PINK_50;
const playerX = 'X';
const playerO = 'O';

export const TicTacToeGame: React.FC = () => {
    const [currentPlayer, setCurrentPlayer] = useState(playerX);
    const [playerIndicatorColor, setPlayerIndicatorColor] = useState(playerXColor);
    const [cellValues, setCellValues] = useState<Array<string | null>>(initialCellValues);
    const [isGameFinished, setIsGameFinishedStatus] = useState(false);
    const [isGameWin, setIsGameWinStatus] = useState(false);

    useEffect(() => {
        const winner = calculateWinner(cellValues);

        if (winner) {
            Alert.alert(`Congratulations ${winner}. You are the winner !`);
            setIsGameFinishedStatus(true);
            setIsGameWinStatus(true);
            setPlayerIndicatorColor(winner === playerX ? playerXColor : playerYColor);
        } else {
            const remainingSpace = cellValues.filter(element => element === null).length;
            if (remainingSpace === 0) {
                Alert.alert('Game Over !');
                setIsGameFinishedStatus(true);
                setIsGameWinStatus(false);
            }
        }
    }, [cellValues]);

    const handleCellPress = (index: number) => {
        let tempArray = [...cellValues];
        if (tempArray[index] === null) {
            tempArray[index] = currentPlayer;
            setCellValues(tempArray);
            setCurrentPlayer(currentPlayer === playerX ? playerO : playerX);
            setPlayerIndicatorColor(currentPlayer === playerX ? playerYColor : playerXColor);
        }
    };

    const calculateWinner = (inputArray: Array<String | null>) => {
        for (const row of winingRowIndexes) {
            const [a, b, c] = row;
            if (inputArray[a]) {
                const isAllInputsEqual = isAllEqual([inputArray[a], inputArray[b], inputArray[c]]);
                if (isAllInputsEqual) {
                    return inputArray[a];
                }
            }
        }
        return null;
    };

    const resetGame = () => {
        setCellValues(initialCellValues);
        setCurrentPlayer(playerX);
        setIsGameFinishedStatus(false);
        setIsGameWinStatus(false);
        setPlayerIndicatorColor(playerXColor);
    };

    const Player = () => {
        return (
            <Image
                testID="playerIcon"
                style={[styles.playerIcon, { tintColor: playerIndicatorColor }]}
                source={require('../../assets/images/icons/player-icon.png')}
            />
        );
    };

    const Winner = () => {
        return (
            <BlinkComponent duration={500}>
                <Player />
            </BlinkComponent>
        );
    };

    type boardCellProps = {
        cellIndex: number;
    };

    const BoardCell = ({ cellIndex }: boardCellProps) => {
        const cellValueColor = cellValues[cellIndex] === playerX ? playerXColor : playerYColor;

        return (
            <TouchableWithoutFeedback disabled={isGameFinished} onPress={() => handleCellPress(cellIndex)}>
                <View style={[styles.cellContainer]}>
                    <Text style={[styles.cellText, { color: cellValueColor }]}>{cellValues[cellIndex]}</Text>
                </View>
            </TouchableWithoutFeedback>
        );
    };

    const GameBoard = () => {
        return (
            <View style={styles.boardContainer}>
                <GridBackground
                    cellWidth={cellWidth}
                    cellContainerWidth={cellContainerWidth}
                    numberOfHorizontalLines={2}
                    numberOfVerticalLines={2}
                />
                <View style={styles.boardCellContainer}>
                    {[...Array(numberOfBoardCells)].map((cell, index) => (
                        <BoardCell key={index} cellIndex={index} />
                    ))}
                </View>
            </View>
        );
    };

    const ResetButton = () => {
        return (
            <TouchableOpacity style={styles.resetButton} onPress={resetGame}>
                <Image style={styles.resetButtonIcon} source={require('../../assets/images/icons/reset-icon.png')} />
            </TouchableOpacity>
        );
    };

    return (
        <View style={styles.container}>
            <View style={styles.playerContainer}>{(isGameWin && <Winner />) || <Player />}</View>
            <View style={styles.gameContainer}>
                <GameBoard />
            </View>
            <View style={styles.resetButtonContainer}>
                <ResetButton />
            </View>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        width: '100%',
        height: '100%',
        backgroundColor: COLORS.BLACK,
        alignItems: 'center',
    },
    gameContainer: { flex: 3, justifyContent: 'center', alignItems: 'center' },
    boardContainer: {
        justifyContent: 'center',
        alignItems: 'center',
    },
    boardCellContainer: {
        flexDirection: 'row',
        flexWrap: 'wrap',
        justifyContent: 'center',
        alignItems: 'center',
        position: 'absolute',
    },
    cellContainer: {
        width: cellWidth,
        height: cellWidth,
        justifyContent: 'center',
        alignItems: 'center',
    },
    cellText: {
        fontSize: 50,
        fontWeight: '800',
    },
    playerContainer: { flex: 1, justifyContent: 'center' },
    playerIcon: { width: 80, height: 80 },
    resetButtonContainer: { flex: 1, width: '100%' },
    resetButton: { alignItems: 'flex-end', marginRight: 20 },
    resetButtonIcon: { width: 80, height: 80 },
});
