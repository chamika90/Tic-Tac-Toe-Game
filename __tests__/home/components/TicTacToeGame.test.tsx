import { render } from '@testing-library/react-native';
import * as React from 'react';

import { TicTacToeGame } from '../../../src/home/components/TicTacToeGame';

describe('Tic Tac Toe Game ', () => {
    it('renders game screen', () => {
        const { toJSON } = render(<TicTacToeGame />);
        expect(toJSON()).toMatchSnapshot();
    });

    it('should render player component', () => {
        const { getByTestId } = render(<TicTacToeGame />);
        expect(getByTestId('playerIcon')).toBeTruthy();
    });
});
