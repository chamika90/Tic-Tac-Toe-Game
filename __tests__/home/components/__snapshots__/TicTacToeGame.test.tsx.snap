// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tic Tac Toe Game  renders game screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/images/icons/player-icon.png",
        }
      }
      style={
        Array [
          Object {
            "height": 80,
            "width": 80,
          },
          Object {
            "tintColor": "#8CE6F8",
          },
        ]
      }
      testID="playerIcon"
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 3,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "height": 642.8571428571429,
            "width": 642.8571428571429,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "position": "absolute",
            }
          }
          testID="horizontalLineContainer"
        >
          <View
            style={
              Array [
                Object {
                  "height": 642.8571428571429,
                  "marginLeft": 214.28571428571428,
                },
                Object {
                  "backgroundColor": "#fff",
                  "width": 1,
                },
              ]
            }
            testID="horizontalLine"
          />
          <View
            style={
              Array [
                Object {
                  "height": 642.8571428571429,
                  "marginLeft": 214.28571428571428,
                },
                Object {
                  "backgroundColor": "#fff",
                  "width": 1,
                },
              ]
            }
            testID="horizontalLine"
          />
        </View>
        <View
          style={
            Object {
              "flexDirection": "column",
              "position": "absolute",
            }
          }
          testID="verticalLineContainer"
        >
          <View
            style={
              Array [
                Object {
                  "marginTop": 214.28571428571428,
                  "width": 642.8571428571429,
                },
                Object {
                  "backgroundColor": "#fff",
                  "height": 1,
                },
              ]
            }
            testID="verticalLine"
          />
          <View
            style={
              Array [
                Object {
                  "marginTop": 214.28571428571428,
                  "width": 642.8571428571429,
                },
                Object {
                  "backgroundColor": "#fff",
                  "height": 1,
                },
              ]
            }
            testID="verticalLine"
          />
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "center",
            "position": "absolute",
          }
        }
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 214.28571428571428,
                "justifyContent": "center",
                "width": 214.28571428571428,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 50,
                  "fontWeight": "800",
                },
                Object {
                  "color": "#F3A6FA",
                },
              ]
            }
          />
        </View>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 214.28571428571428,
                "justifyContent": "center",
                "width": 214.28571428571428,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 50,
                  "fontWeight": "800",
                },
                Object {
                  "color": "#F3A6FA",
                },
              ]
            }
          />
        </View>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 214.28571428571428,
                "justifyContent": "center",
                "width": 214.28571428571428,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 50,
                  "fontWeight": "800",
                },
                Object {
                  "color": "#F3A6FA",
                },
              ]
            }
          />
        </View>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 214.28571428571428,
                "justifyContent": "center",
                "width": 214.28571428571428,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 50,
                  "fontWeight": "800",
                },
                Object {
                  "color": "#F3A6FA",
                },
              ]
            }
          />
        </View>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 214.28571428571428,
                "justifyContent": "center",
                "width": 214.28571428571428,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 50,
                  "fontWeight": "800",
                },
                Object {
                  "color": "#F3A6FA",
                },
              ]
            }
          />
        </View>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 214.28571428571428,
                "justifyContent": "center",
                "width": 214.28571428571428,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 50,
                  "fontWeight": "800",
                },
                Object {
                  "color": "#F3A6FA",
                },
              ]
            }
          />
        </View>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 214.28571428571428,
                "justifyContent": "center",
                "width": 214.28571428571428,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 50,
                  "fontWeight": "800",
                },
                Object {
                  "color": "#F3A6FA",
                },
              ]
            }
          />
        </View>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 214.28571428571428,
                "justifyContent": "center",
                "width": 214.28571428571428,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 50,
                  "fontWeight": "800",
                },
                Object {
                  "color": "#F3A6FA",
                },
              ]
            }
          />
        </View>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 214.28571428571428,
                "justifyContent": "center",
                "width": 214.28571428571428,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 50,
                  "fontWeight": "800",
                },
                Object {
                  "color": "#F3A6FA",
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "flex-end",
          "marginRight": 20,
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/images/icons/reset-icon.png",
          }
        }
        style={
          Object {
            "height": 80,
            "width": 80,
          }
        }
      />
    </View>
  </View>
</View>
`;
